python/ray/experimental/ddp/ddp.py

```diff
@@ -0,0 +1,541 @@
+        """
+        Gathers all weights across different layers.
+
+        Args:
+            updates: A tuple of any number of updated weights.
+        """
+        return updates
+
+    def get_grad_to_reduce(
+        self, grad: Tuple[torch.Tensor, torch.Tensor]
+    ) -> torch.Tensor:
+        """
+        When an allreduce binds a class method output with `num_returns > 1`,
+        an error is thrown. This is a workaround.
+        See: https://github.com/ray-project/ray/issues/48522
+        """
+        _, reduce_grad = grad
+        return reduce_grad
+
+
+def generate_input_output(config: Config) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Generate input `x` and output `y` for training."""
+    layer_size = config.layer_size
+    num_actors = config.num_actors
+    dtype = config.dtype
+
+    shape = (num_actors * layer_size, layer_size)
+    numel = shape[0] * shape[1]
+
+    x = torch.arange(numel, dtype=dtype, requires_grad=True) * 0.1
+    x = x.reshape(shape)
+    y = torch.arange(numel, dtype=dtype) * 0.1
+    y = y.reshape(shape)
+
+    return x, y
+
+
+def run_torch(config):
+    """Run pytorch without DDP and return the weights."""
+    device = "cuda:0"
+    model = LayeredModel(
+        config.layer_size, config.num_layers, device, config.dtype, config.lr
+    )
+    criterion = model.criterion
+    optimizer = optim.SGD(model.parameters(), lr=model.lr)
+
+    torch_weights = []
+    elapses = []
+    for i in range(config.it):
+        x, y = generate_input_output(config)
+        x = x.to(device)
+        y = y.to(device)
+        start = time.perf_counter()
+        optimizer.zero_grad()
+        pred: torch.Tensor = model(x)
+        loss: torch.Tensor = criterion(pred, y)
+        loss.backward()
+        optimizer.step()
+        end = time.perf_counter()
+
+        cur_iter_weights = []
+        for i in range(0, len(model.layers), 2):
+            layer: torch.nn.Linear = model.layers[i]
+            cur_iter_weights.append(torch.clone(layer.weight))
+        torch_weights.append(cur_iter_weights)
+
+        elapse = end - start
+        elapses.append(elapse)
+
+    print_elapses(elapses, "torch")
+
+    return torch_weights
+
+
+def run_torch_ddp(config: Config):
+    """Run DDP with `torch.distributed`."""
+    n_gpus = torch.cuda.device_count()
+    assert (
+        n_gpus >= config.num_actors
+    ), f"Requires at least {config.num_actors} GPUs to run, but got {n_gpus}"
+    world_size = config.num_actors
+    mp.set_start_method("spawn", force=True)
+    with mp.Manager() as manager:
+        return_dict = manager.dict()
+        mp.spawn(
+            torch_ddp_per_process,
+            args=(world_size, return_dict, config),
+            nprocs=world_size,
+            join=True,
+        )
+        torch_ddp_weights = get_torch_ddp_weights_per_device(return_dict, world_size)
+        return torch_ddp_weights
+
+
+def torch_ddp_per_process(rank, world_size, return_dict, config: Config):
```

Comment:
Nit: Could you add the missing type hints? There are other functions missing too.

Reply:
Nit: Rename to `run_torch_ddp_per_process`.

Reply:
Added type annotations.